### PR TITLE
Add support for Custom entitlements files

### PR DESCRIPTION
**Pull request by yyosifov**
Adding Entitlements support:

- We now create by default an entitlements file for you
- If you add an `app.entitlements` file in the App_Resources folder it will be added and merged
- If you have plugins with `app.entitlements` in the `platforms/ios` folder they will be considered and merged
- If you haven't set the `CODE_SIGN_ENTITLEMENTS` property in xcconfig, we set it by default to the default entitlements file we produce.

Small refactoring of the XCConfig read property functionality - extracted, fixed several issues and added tests to cover all.

### CODE REVIEW

File: test/ios-entitlements-service.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	};
+
+	let injector: IInjector,
+		platformsData: any,
+		projectData: IProjectData,
+		fs: IFileSystem,
+		iOSEntitlementsService: IOSEntitlementsService,
+		destinationFilePath: string;
+
+	beforeEach(function() {
```
Comment (rosen-vladimirov) [inline test/ios-entitlements-service.ts]: `beforeEach( () => {`

File: test/ios-entitlements-service.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+		iOSEntitlementsService = injector.resolve("iOSEntitlementsService");
+		destinationFilePath = iOSEntitlementsService.getPlatformsEntitlementsPath(projectData);
+	});
+
+	describe("Ensure paths constructed are correct", () => {
+		it("Ensure destination entitlements relative path is calculated correctly.", () => {
+			const expected = "testApp/testApp.entitlements";
+			let actual = iOSEntitlementsService.getPlatformsEntitlementsRelativePath(projectData);
+			assert.equal(actual, expected);
```
Comment (rosen-vladimirov) [inline test/ios-entitlements-service.ts]: `assert.deepEqual` maybe?

File: test/ios-entitlements-service.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    <key>nameKey</key>
+    <string>appResources</string>
+  </dict>
+</plist>`;
+
+		function assertContent(actual: string, expected: string) {
+			let strip = (x: string) => {
+				return x.replace(EOL, '').trim();
+			};
+			assert.equal(strip(actual), strip(expected));
```
Comment (rosen-vladimirov) [inline test/ios-entitlements-service.ts]: `deepEqual` maybe

File: test/ios-project-service.ts [hunk 4]
```diff
@@ -720,8 +744,115 @@ describe("iOS Project Service Signing", () => {
...
+
+			assert.isTrue(fs.exists(destinationFilePath), 'Target build xcconfig is missing for release: ' + release);
+			let expected = {
+				'CODE_SIGN_ENTITLEMENTS': iOSEntitlementsService.getPlatformsEntitlementsRelativePath(projectData)
+			};
+			assertPropertyValues(expected, destinationFilePath, testInjector);
+		}
+	});
+
+	it("The user specified entitlements property takes percedence", async () => {
```
Comment (rosen-vladimirov) [inline test/ios-project-service.ts]: `percedence` -> `precedence` maybe

File: test/xcconfig-service.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+		testInjector.register('xCConfigService', XCConfigService);
+
+		return testInjector;
+	};
+
+	const assertPropertyValues = (expected: any, injector: IInjector) => {
+		let service = getXCConfigService(injector);
+		_.forOwn(expected, (value, key) => {
+			let actual = service.readPropertyValue("any", key);
+			assert.equal(actual, value);
```
Comment (rosen-vladimirov) [inline test/xcconfig-service.ts]: deepEqual

File: lib/constants.ts [hunk 6]
```diff
@@ -16,6 +16,9 @@ export const TEST_RUNNER_NAME = "nativescript-unit-test-runner";
 export const LIVESYNC_EXCLUDED_FILE_PATTERNS = ["**/*.js.map", "**/*.ts"];
 export const XML_FILE_EXTENSION = ".xml";
 export const PLATFORMS_DIR_NAME = "platforms";
+export const IOS_PLATFORM_NAME = "ios";
+export const IOS_PLATFORM_NORMALIZED_NAME = "iOS";
```
Comment (yyosifov) [inline lib/constants.ts]: I'll do it, although there are already some constants in ios-project-service for "ios" and it turned out my code pretty ugly from:
`constants.IOS_PLATFORM_NORMALIZED_NAME`
to:
`this.$mobileHelper.normalizePlatformName(this.$devicePlatformsConstants.iOS)`

File: lib/services/ios-project-service.ts [hunk 7]
```diff
@@ -40,9 +42,11 @@ export class IOSProjectService extends projectServiceBaseLib.PlatformProjectServ
 		private $pluginVariablesService: IPluginVariablesService,
 		private $xcprojService: IXcprojService,
 		private $iOSProvisionService: IOSProvisionService,
-		private $sysInfo: ISysInfo,
 		private $pbxprojDomXcode: IPbxprojDomXcode,
-		private $xcode: IXcode) {
+		private $xcode: IXcode,
+		private $iOSEntitlementsService: IOSEntitlementsService,
+		private $sysInfo: ISysInfo,
+		private xCConfigService: XCConfigService) {
```
Comment (yyosifov) [inline lib/services/ios-project-service.ts]: Fixed. What does this `$` stand for anyway?

File: test/ios-entitlements-service.ts (hunk 2 above)
Comment (yyosifov) [inline test/ios-entitlements-service.ts]: That's a single string (see `expected` value). Do we need deepEqual in this case?

File: test/ios-entitlements-service.ts (hunk 3 above)
Comment (yyosifov) [inline test/ios-entitlements-service.ts]: again, these are string? Do we need it here?

File: test/xcconfig-service.ts (hunk 5 above)
Comment (yyosifov) [inline test/xcconfig-service.ts]: These are strings. Do we need it here?

Comment (rosen-vladimirov) [inline test/xcconfig-service.ts]: nope, sorry, my mistake, equal is fine

**Comment by rynop** [comment]
This is great guys, thanks. Is this [documented](https://docs.nativescript.org/) anywhere?  I think this function is common across many plugins, and some go as far as even [duplicating this logic](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/blob/master/src/scripts/entitlements-before-prepare.js#L12).

If it is not documented, I'd be happy to write it up.  Just looking for guidance on where in the current docs to put it.

**Pull request merged**